docs/proposals/device-management-enhance.md

```diff
@@ -0,0 +1,1069 @@
+	...
+	// Protocol configuration for bluetooth
+	// +optional
+	Bluetooth *ProtocolConfigBluetooth `json:"bluetooth,omitempty"`
+	// Configuration for protocol common part
+	// +optional
+	Common *ProtocolConfigCommon `json:"common,omitempty"`
+	// Configuration for customized protocol
+	// +optional
+	CustomizedProtocol *ProtocolConfigCustomized `json:"customizedProtocol,omitempty"`
+}
+
+type ProtocolConfigModbus struct {
+	// Required. 0-255
+	SlaveID int64 `json:"slaveID,omitempty"`
+}
+
+// Only one of COM or TCP may be specified.
+type ProtocolConfigCommon struct {
+	// +optional
+	COM *ProtocolConfigCOM `json:"com,omitempty"`
+	// +optional
+	TCP *ProtocolConfigTCP `json:"tcp,omitempty"`
+	// Communication type, like tcp client, tcp server or COM
+	// +optional
+	RTU *ProtocolConfigModbusRTU `json:"rtu,omitempty"`
+	CommType string `json:"commType,omitempty"`
+	// Reconnection timeout
+	// +optional
+	TCP *ProtocolConfigModbusTCP `json:"tcp,omitempty"`
+	ReconnTimeout int64 `json:"reconnTimeout,omitempty"`
+	// Reconnecting retry times
+	// +optional
+	ReconnRetryTimes int64 `json:"reconnRetryTimes,omitempty"`
+	// Define timeout of mapper collect from device.
+	// +optional
+	CollectTimeout int64 `json:"collectTimeout,omitempty"`
+	// Define retry times of mapper will collect from device.
+	// +optional
+	CollectRetryTimes int64 `json:"collectRetryTimes,omitempty"`
+	// Define collect type, sync or async.
+	// +optional
+	CollectType string `json:"collectType,omitempty"`
+	// Customized values for provided protocol
+	// +optional
+	CustomizedValues *CustomizedValue `json:"customizedValues,omitempty"`
+}
+
+type ProtocolConfigTCP struct {
+	// Required.
+	IP string `json:"ip,omitempty"`
+	// Required.
+	Port int64 `json:"port,omitempty"`
+}
+
+type ProtocolConfigCOM struct {
+	...
+	// Required.
+	SerialPort string `json:"serialPort,omitempty"`
+	// Required. BaudRate 115200|57600|38400|19200|9600|4800|2400|1800|1200|600|300|200|150|134|110|75|50
+	Parity string `json:"parity,omitempty"`
+	// Required. Bit that stops 1|2
+	StopBits int64 `json:"stopBits,omitempty"`
+}
+
+type ProtocolConfigCustomized struct {
+	// Unique protocol name
+	// Required.
+	ProtocolName string `json:"protocolName,omitempty"`
+	// Any config data
+	// +optional
+	ConfigData *CustomizedValue `json:"configData,omitempty"`
+}
+
+// DeviceData reports the device's time-series data to edge MQTT broker.
+// These data should not be processed by edgecore. Instead, they can be process by
+// third-party data-processing apps like EMQX kuiper.
+type DeviceData struct {
+	// Required: A list of data properties, which are not required to be processed by edgecore
+	DataProperties []DataProperty `json:"dataProperties,omitempty"`
+	// Topic used by mapper, all data collected from dataProperties
+	// should be published to this topic,
+	// the default value is $ke/events/device/+/data/update
+	// +optional
+	DataTopic string `json:"dataTopic,omitempty"`
```

Comment:
how does third-party data-processing apps like EMQX kuiper get the topic?

Reply:
third-party can get this info by using configMap generated by KubeEdge, or they can use Device API query this info and add it to their own configMap